src/Analyser/NodeScopeResolver.php

```diff
@@ -1955,12 +1955,7 @@ static function (?Type $offsetType, Type $valueType, bool $optional) use (&$arra
 
 			if (isset($functionReflection) && $functionReflection->getName() === 'shuffle') {
 				$arrayArg = $expr->getArgs()[0]->value;
-				$arrayArgType = $scope->getType($arrayArg);
-
-				if ($arrayArgType instanceof ConstantArrayType) {
-					$arrayArgType = $arrayArgType->getValuesArray()->generalizeToArray();
-				}
-
+				$arrayArgType = $scope->getType($arrayArg)->shuffleArray();
 				$scope = $scope->assignExpression($arrayArg, $arrayArgType, $arrayArgType);
```

Comment:
based on your last comment in the popArray() PR - not sure what should be done with the native type here. also getting it and apply the method? definitely more tests for that are needed :)

Reply:
The last arg should be `$scope->getNativeType($arrayArg)->shuffleArray()`.

You can test it with assertNativeType.

It isn't always less precise, there are situations when we know ConstantArrayType as native type too, for example after `$a = [...];`

But assertNativeType should be different when the argument type comes from PHPDoc (array shape).